torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,174 @@
+import torch
+import contextlib
+from typing import Iterator
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.utils._python_dispatch import enable_python_mode
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+    guard = torch._C._DisableTorchDispatch()  # type: ignore[attr-defined]
+    try:
+        yield
+    finally:
+        del guard
+
+def check_metadata_consistency(wrapper_tensor):
+    if not isinstance(wrapper_tensor, CompositeCompliantTensor):
+        return
+    elem = wrapper_tensor.elem
+    if wrapper_tensor.shape != elem.shape:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "shape of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+    if wrapper_tensor.dtype != elem.dtype:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "dtype of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+
+# This is a bit fragile because this needs to be updated whenever a new view op
+# is added. We probably want some sort of introspection to tell that if a
+# torch.* op is a view or not.
+def is_view_fn(func):
+    return func.__name__ in {
+        'as_strided',
+        'detach',
+        'diagonal',
+        'expand',
+        'expand_as',
+        'movedim',
+        'narrow',
+        'permute',
+        'select',
+        'squeeze',
+        'transpose',
+        't',
+        'real',
+        'imag',
+        'view_as_real',
+        'view_as_complex',
+        'unflatten',
+        'unfold',
+        'unsqueeze',
+        'view',
+        'view_as',
+        'unbind',
+        'split',
+        'split_with_sizes',
+        'vsplit',
+        'hsplit',
+        'tensor_split',
+        'chunk',
+        'swapaxes',
+        'slice',
+        '_reshape_alias',
+        '_unsafe_view',
+        '_conj',
+    }
+
+def is_inplace_view_fn(func):
+    return func.__name__ in {
+        'squeeze_',
+        'unsqueeze_',
+        'transpose_',
+        't_',
+    }
+
+class CompositeCompliantTensor(torch.Tensor):
+    elem: torch.Tensor
+
+    __slots__ = ['elem']
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        # The storage of CompositeCompliantTensor should never be used directly
+        # by a CompositeImplicitAutograd operation; if the CompositeImplicitAutograd
+        # operator attempts to read from the storage without dispatching then it'll
+        # raise a RuntimeError due to it being a meta storage.
```

Comment:
will this raise whatever default error gets thrown by `storage_access_should_throw`? We probably want this test to give a more actionable error message.

Reply:
The test catches the error, adds some context with actionable information, and then rethrows the error with additional context. So it looks like the following:
```
================================================= FAILURES =================================================
___________________ TestCommonCPU.test_composite_compliance_linalg_cholesky_cpu_float32 ____________________
Traceback (most recent call last):
  File "/raid/rzou/pt/workspace/torch/testing/_internal/composite_compliance.py", line 165, in _check_compos
ite_compliance
    op(*args, **kwargs)
  File "/raid/rzou/pt/workspace/torch/testing/_internal/common_methods_invocations.py", line 627, in __call_
_
    return self.op(*args, **kwargs)
  File "/raid/rzou/pt/workspace/torch/_tensor.py", line 1064, in __torch_function__
    ret = func(*args, **kwargs)
RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a
 lazy allocation, so you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/raid/rzou/pt/workspace/test/test_ops.py", line 515, in test_composite_compliance
    _check_composite_compliance(op, args, kwargs)
  File "/raid/rzou/pt/workspace/torch/testing/_internal/composite_compliance.py", line 167, in _check_compos
ite_compliance
    raise RuntimeError(f"CompositeImplicitAutograd compilance check failed with "
RuntimeError: CompositeImplicitAutograd compilance check failed with the following error. If you are adding
an OpInfo of an existing operator, please feel free to skip this test because the problem was pre-existing a
nd file an issue. Otherwise, if you added a new operator, please read through the CompositeImplicitAutograd
Compliance section in aten/src/ATen/native/README.md for how to resolve this. Got error message: The tensor
has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, so you
will need to call mutable_data() or raw_mutable_data() to actually allocate memory.

```